### ISSUE TITLE
quotes rule support for backtick/backquote option

### ISSUE THREAD

**Issue by zaggino** [eslint/eslint#2153]
Hi guys,
in ES6 it's legal to use apostrophes for strings (template strings):
```
let a = `string`;
```
Is it a bad idea to use it generally in the code instead of single and double quotes?
If not, could the `quotes` rule support this as an option?

**Comment by michaelficarra** [comment]
Single quote is apostrophe. Those are backticks. Agreed there should be an option for it.

**Comment by nzakas** [comment]
Yeah, probably a good idea.

**Comment by doberkofler** [comment]
@zaggino If I get this right, this option would allow to (for example) specify that all strings must be backticks and I'm not sure if this is what people really want. I would imagine that the use of backticks would be additional (when needed) to the use of single or double quotes and not exclusive.

**Comment by zaggino** [comment]
Well I wouldn't want it to be default or something, but unless there's a good reason, I'd like to be able to force backticks in my projects, why not? It'd be an optional setting, not a forced one.

**Comment by nzakas** [comment]
@doberkofler there are two scenarios:

1. Enforce backticks for all strings (disallow single and double quotes)
2. Enforce single or double quotes AND ALSO allow backticks

**Comment by doberkofler** [comment]
@zaggino What @nzakas described (including scenario 2) is exactly what I think needs to be supported.
I see the following 5 combinations:
- Only use single quotes
- Only use double quotes
- Only use backticks
- Only use single quotes or backticks
- Only use double quotes or backticks

**Comment by vlinder** [comment]
@nzakas As @doberkofler commented. The desired use cases would force single or double quotes and still allow for backticks. But one would not want to force backticks? I cannot see this implemented in the source. This bug should probably be reopened?

**Comment by nzakas** [comment]
You can require backticks by using `backtick`. See: http://eslint.org/docs/rules/quotes

**Comment by doberkofler** [comment]
@nzakas I would imagine that most people (like me) will migrate from single or double quotes to additionally using backticks where needed and only later to eventually only using backticks. That's the reason why I suggested that eslint should support the 5 options I've described.

**Comment by vlinder** [comment]
Exactly like @doberkofler wrote, force one of double or single and also allow for backticks (and still being able to use "avoid-escape").

Something like this:
```
  "quotes": [1, "single", "backtick", "avoid-escape"],
```

**Comment by nzakas** [comment]
@doberkofler you can already do that with the three options available. I'm not convinced we need an explicit flag to disallow template literals in this rule, especially since there is a feature flag for template literals that you can just turn off.

If you want to discuss more, please open a separate issue. This one is closed because it has been addressed.

**Comment by granteagon** [comment]
Looks like template strings aren't allowed.  I must be missing something:

```es6
const Icon = prop =>
  <i className="`${ prop.icon.className }`" />
```

I get a 'no-unused-vars' "prop" is defined but never used.

My config:
```json
{
  "env": {
    "browser": true,
    "node": true,
    "mocha": true
  },
  "plugins": [
    "react"
  ],
  "settings": {
    "react": {
      "pragma": "React",
      "version": "15.0"
    }
  },
  "parserOptions": {
    "ecmaVersion": 7,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true,
      "modules": true
    }
  },
  "extends": ["eslint:recommended", "plugin:react/recommended"],
  "rules": {
    "quotes": ["error", "single"],
    "strict": [2, "never"],
    "no-console": 1,
    "no-alert": 1,
    "no-debugger": 1,
    "no-cond-assign": 2,
    "no-constant-condition": 2,
    "no-control-regex": 2,
    "no-dupe-args": 2,
    "no-dupe-keys": 2,
    "no-duplicate-case": 2,
    "no-empty": 2,
    "no-empty-character-class": 2,
    "no-ex-assign": 2,
    "no-extra-boolean-cast": 2,
    "no-extra-parens": 1,
    "no-extra-semi": 2,
    "no-func-assign": 2,
    "no-inner-declarations": 2,
    "no-invalid-regexp": 2,
    "no-irregular-whitespace": 2,
    "no-negated-in-lhs": 2,
    "no-obj-calls": 2,
    "no-regex-spaces": 2,
    "no-sparse-arrays": 2,
    "no-unexpected-multiline": 2,
    "no-unreachable": 2,
    "use-isnan": 2,
    "valid-jsdoc": 2,
    "valid-typeof": 2,
    "accessor-pairs": 2,
    "array-callback-return": 2,
    "block-scoped-var": 2,
    "complexity": [1, 5],
    "consistent-return": 2,
    "curly": 2,
    "default-case": 2,
    "dot-location": 0,
    "dot-notation": 2,
    "eqeqeq": 2,
    "guard-for-in": 2,
    "no-caller": 2,
    "no-case-declarations": 2,
    "no-div-regex": 2,
    "no-else-return": 2,
    "no-empty-function": [2, {"allow": []}],
    "no-empty-pattern": 2,
    "no-eq-null": 2,
    "no-eval": 2,
    "no-extend-native": 2,
    "no-extra-bind": 2,
    "no-extra-label": 2,
    "no-fallthrough": 2,
    "no-floating-decimal": 2,
    "no-implicit-coercion": 2,
    "no-implicit-globals": 2,
    "no-implied-eval": 2,
    "no-invalid-this": 2,
    "no-iterator": 2,
    "no-labels": 2,
    "no-lone-blocks": 2,
    "no-loop-func": 2,
    "no-magic-numbers": 0,
    "no-multi-spaces": 2,
    "no-multi-str": 2,
    "no-native-reassign": 2,
    "no-new": 2,
    "no-new-func": 2,
    "no-new-wrappers": 2,
    "no-octal": 2,
    "no-octal-escape": 2,
    "no-param-reassign": 2,
    "no-process-env": 2,
    "no-proto": 2,
    "no-redeclare": 2,
    "no-return-assign": 2,
    "no-script-url": 2,
    "no-self-assign": 2,
    "no-self-compare": 2,
    "no-sequences": 2,
    "no-throw-literal": 2,
    "no-unmodified-loop-condition": 2,
    "no-unused-expressions": 2,
    "no-unused-labels": 2,
    "no-useless-call": 2,
    "no-useless-concat": 2,
    "no-void": 2,
    "no-warning-comments": [1, { "terms": ["todo", "fixme", "xxx"], "location": "start" }],
    "no-with": 2,
    "radix": 2,
    "vars-on-top": 2,
    "wrap-iife": 2,
    "yoda": 2,
    "no-catch-shadow": 2,
    "no-delete-var": 2,
    "no-label-var": 2,
    "no-shadow": 2,
    "no-shadow-restricted-names": 2,
    "no-undef": 2,
    "no-undef-init": 2,
    "no-unused-vars": 1,
    "no-use-before-define": 2,
    "callback-return": 2,
    "handle-callback-err": 2,
    "no-mixed-requires": 2,
    "no-new-require": 2,
    "no-path-concat": 2,
    "no-process-exit": 2,
    "no-sync": 2,
    "block-spacing": [2, "always"],
    "brace-style": [2, "1tbs", { "allowSingleLine": true }],
    "camelcase": [2, {"properties": "always"}],
    "comma-spacing": [2, {"before": false, "after": true}]
  }
}
```

**Comment by mysticatea** [comment]
@granteagon it's just a string literal (it's enclosed by double quotes), not a template literal.

**Comment by granteagon** [comment]
@mysticatea Okay, but it still throws 'no-unused-vars' validation because it thinks prop is unused.  That's the problem I'm trying to solve.

![no-unused-vars](https://s3.amazonaws.com/f.cl.ly/items/210L0e3Y3S0x132l1S1u/Image%202016.07.25%208%3A38%3A06%20AM.png?v=ee15e64c)

**Comment by mysticatea** [comment]
@granteagon Yes, the `prop` is not used. So it's correct message. 

    "`${ prop.icon.className }`"

is a string literal (it's enclosed by double quotes), just a text, so it's not using `prop`.

Maybe your intention is:

```js
const Icon = prop =>
  <i className={ prop.icon.className } />
```

**Comment by granteagon** [comment]
@mysticatea, got it.  What I was trying to do was:

```es6
className={ `Icon ${ prop.icon.className }` }
```

Got my syntax confused.  Thanks!
